vertx-grpc-server/src/main/java/io/vertx/grpc/server/auth/impl/GrpcJWTAuthenticationHandlerImpl.java

```diff
@@ -0,0 +1,137 @@
+package io.vertx.grpc.server.auth.impl;
+
+import io.vertx.core.Future;
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.ext.auth.User;
+import io.vertx.ext.auth.authentication.AuthenticationProvider;
+import io.vertx.ext.auth.authentication.TokenCredentials;
+import io.vertx.ext.auth.jwt.JWTAuth;
+import io.vertx.grpc.common.GrpcStatus;
+import io.vertx.grpc.server.GrpcException;
+import io.vertx.grpc.server.auth.GrpcJWTAuthenticationHandler;
+
+public class GrpcJWTAuthenticationHandlerImpl implements GrpcJWTAuthenticationHandler {
+
+  static final GrpcException UNAUTHENTICATED = new GrpcException(
+    GrpcStatus.UNAUTHENTICATED);
+  static final GrpcException UNKNOWN = new GrpcException(GrpcStatus.UNKNOWN);
+  static final GrpcException UNIMPLEMENTED = new GrpcException(
+    GrpcStatus.UNIMPLEMENTED);
+
+  public enum Type {
+    BASIC("Basic"), DIGEST("Digest"), BEARER("Bearer"),
+    // these have no known implementation
+    HOBA("HOBA"), MUTUAL("Mutual"), NEGOTIATE("Negotiate"), OAUTH(
+      "OAuth"), SCRAM_SHA_1("SCRAM-SHA-1"), SCRAM_SHA_256("SCRAM-SHA-256");
+
+    private final String label;
+
+    Type(String label) {
+      this.label = label;
+    }
+
+    public boolean is(String other) {
+      return label.equalsIgnoreCase(other);
+    }
+
+    @Override
+    public String toString() {
+      return label;
+    }
+  }
+
+  private final AuthenticationProvider authProvider;
+  private final Type type;
+  private final String realm;
+
+  public GrpcJWTAuthenticationHandlerImpl(JWTAuth authProvider, String realm) {
+    this(authProvider, Type.BEARER, realm);
+  }
+
+  public GrpcJWTAuthenticationHandlerImpl(AuthenticationProvider authProvider,
+    Type type, String realm) {
+    this.authProvider = authProvider;
+    this.type = type;
+    this.realm = realm == null ? null
+      : realm
+        // escape quotes
+        .replaceAll("\"", "\\\"");
+
+    if (this.realm != null &&
+      (this.realm.indexOf('\r') != -1 || this.realm.indexOf('\n') != -1)) {
+      throw new IllegalArgumentException(
+        "Not allowed [\\r|\\n] characters detected on realm name");
+    }
+  }
+
+  private Future<String> parseAuthorization(HttpServerRequest req,
```

Comment:
we should use a result here and avoid relying on exception to communicate an authentication failure (and thus avoid GrpcException), the result should clearly be an enum like result and the gRPC code calling for authentication should set the corresponding status on the HTTP/2 response

Reply:
I agree. Failures should only be used for unexpected states. I started rewriting this to use an enum as suggested:

```java
     return Future.succeededFuture(GrpcAuthenticationResult.UNAUTHENTICATED);
```

I however don't see how I can return the token string in the same method via:

```java
    return Future.succeededFuture(authorization.substring(idx + 1));
```

Do have an example or a pointer for me on how to solve this?

Reply:
I will need to have a look deeper to make an elaborated suggestion here